pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Delete runs through a list of potential candidates and select the ones that can be deleted.
+// Do not run this function unless driver expectations are met.
+func (ctx *rollingUpgradeCtx) Delete() ([]corev1.Pod, error) {
+	if len(ctx.podsToUpgrade) == 0 {
+		return nil, nil
+	}
+
+	// Check if we are not over disruption budget
+	// Upscale is done, we should have the required number of Pods
+	expectedPods := ctx.statefulSets.PodNames()
+	unhealthyPods := len(expectedPods) - len(ctx.healthyPods)
+	maxUnavailable := 1
+	// TODO: use GroupingDefinition
+	if ctx.ES.Spec.UpdateStrategy.ChangeBudget != nil {
+		maxUnavailable = ctx.ES.Spec.UpdateStrategy.ChangeBudget.MaxUnavailable
+	}
+	allowedDeletions := maxUnavailable - unhealthyPods
+	// If maxUnavailable is reached the deletion driver still allows one unhealthy Pod to be restarted.
```

Comment:
I wanted to be conservative here by limiting the number of unhealthy Pods we want to try to delete simultaneously to 1.
The predicate is a safety belt so that only unhealthy Pods can be deleted when `maxUnavailable` is reached.

Reply:
I'm not sure I understand where this "max 1 unhealthy pod restarted" is applied. The first predicate seems to always authorise unhealthy pods to be upgraded?

Reply:
Only 1 will be deleted. This is because whatever the predicate does the loop will immediately break as `allowedDeletions` will be negative as soon as the first one is deleted:

https://github.com/barkbay/cloud-on-k8s/blob/ondelete-master/pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go#L88-L102

Maybe the question would be "should we have something smarter to allow more than one unhealthy Pod the be restarted ?"

Reply:
Got it, missed that.
Let's keep it this way for now I would way, and optimize later if needed.